### PR TITLE
Adding support to IPFS and solidifying upload method

### PR DESCRIPTION
**Pull request by aaitor**
## Description

- Adding support to upload files or messages
- Adding support to IPFS
- Exception catching

## Is this PR related with an open issue?

Related to Issue #57 #59 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [X] Follows the code style of this project.
- [X] Tests Cover Changes
- [ ] Documentation

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "name": "node-ts",
+  "version": "0.4.4",
   "version": "0.5.0",
```
Comment (r-marques) [inline package.json:4]: this needs to be fixed

File: src/access/access.controller.ts [hunk 2]
```diff
@@ -134,43 +137,64 @@ export class AccessController {
...
-        return { url, password };
+        const url = await this.nvmService.uploadFilecoin(data, fileName);
+        return { url };
+      } else if (backend === 'ipfs') {
+        const url = await this.nvmService.uploadIPFS(data, fileName);
+        return { url };
       }
+    } catch (error) {
+      Logger.error(`Error processing upload: ${error.message}`);
+      throw new HttpException(error.message, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (r-marques) [inline src/access/access.controller.ts:196]: we can remove the code duplication here. the only thing that changes with `encrypt` is `data` and `password` all the rest is the same

File: src/shared/nevermined/nvm.service.ts [hunk 3]
```diff
@@ -236,6 +237,37 @@ export class NeverminedService {
...
+      Logger.error(`Uploading ${filename}: IPFS error ${e}`);
+      throw new InternalServerErrorException(e);
+    }
+  }
+
+  private getIPFSAuthToken(): string | undefined {
+
+    if (!this.config.get('IPFS_PROJECT_ID') || !this.config.get('IPFS_PROJECT_SECRET')) {
+      Logger.warn(`Infura IPFS_PROJECT_ID or IPFS_PROJECT_SECRET are not set - disabling ipfs auth`)      
+      return undefined
```
Comment (r-marques) [inline src/shared/nevermined/nvm.service.ts:263]: I think this should just throw an error with a message saying something like `this node does not support upload to ipfs`

**Review by r-marques — APPROVED** [review]

File: src/shared/nevermined/nvm.service.ts (hunk 3 above)
Comment (aaitor) [reply src/shared/nevermined/nvm.service.ts:263]: this warns but allows you still can use a public ipfs now without project_id or secret

File: src/access/access.controller.ts (hunk 2 above)
Comment (aaitor) [reply src/access/access.controller.ts]: done

File: src/access/access.controller.ts [hunk 4]
```diff
@@ -134,43 +137,64 @@ export class AccessController {
...
+      throw new BadRequestException('No file or message in request');
     }
-    let data = file.buffer;
-    if (uploadData.encrypt) {
-      // generate password
-      Logger.debug(`Uploading with password, filename ${file.originalname}`);
-      const password = crypto.randomBytes(32).toString('base64url');
-      data = Buffer.from(aes_encryption_256(data, password), 'binary');
+    if (backend !== 's3' && backend !== 'filecoin' && backend !== 'ipfs')
+      throw new BadRequestException(`Backend ${backend} not supported`);
```
[the remark this answers is not in the document]
Comment (r-marques) [reply src/access/access.controller.ts]: I think you can use @Query instead of @Param

**Pull request merged**